### PR TITLE
fix/normalize storage paths

### PR DESCRIPTION
**Pull request by d-v-b**
This PR adds some of the path normalization logic from zarr v2 into v3. I largely copied [this function](https://github.com/zarr-developers/zarr-python/blob/ed94877ff69d859f2e9d412e775e2887bc6c71bf/zarr/util.py#L340-L380) as-is with a few tweaks. That function is actually not terribly strict as far as path parsing is concerned, so I'm open to requests that we make it stricter. The main thing to note is that `normalize_path` (the new name for that function) strips leading `/` characters from strings, which ensures that paths are relative.

I wired this function up to `make_store_path`, which previously didn't take `path` as a parameter (now it does). This led to some nice code deletion opportunities in our various creation routines.

Something that always bugged me in zarr v2 was use of `None` as a default value for types where there's already a default value. In the case of strings, the empty string `''` is a perfectly good default value, and so allowing `None` is just noise IMO. I was _tempted_ to try and remove `None` as a valid path argument in this PR, but I held back. If people agree that `None` is a silly default for a stringy value like `path` when we can just use `''` instead, then I can make that change as well.

Fixes https://github.com/zarr-developers/zarr-python/issues/2357

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: src/zarr/storage/common.py [hunk 1]
```diff
@@ -209,37 +214,44 @@ async def make_store_path(
...
-
+    path_normalized = normalize_path(path)
     if isinstance(store_like, StorePath):
         if mode is not None and mode != store_like.store.mode.str:
             _store = store_like.store.with_mode(mode)
             await _store._ensure_open()
-            store_like = StorePath(_store)
-        result = store_like
+            store_like = StorePath(_store, path=store_like.path) / path_normalized
+        result = store_like / path_normalized
```
Comment (jhamman) [inline src/zarr/storage/common.py:223]: looks like `/ path_normalized` is being applied twice here.

File: tests/v3/test_api.py [hunk 2]
```diff
@@ -37,6 +48,20 @@ def test_create_array(memory_store: Store) -> None:
     assert z.chunks == (40,)
 
 
+@pytest.mark.parametrize("path", ["/", "///foo/bar"])
```
Comment (jhamman) [inline tests/v3/test_api.py:51]:
```suggestion
@pytest.mark.parametrize("path", ["foo", "/", "/foo", "///foo/bar"])
```

**Review by jhamman — APPROVED** [review]
Good stuff @d-v-b - I'm glad we're moving `path` into `make_store_path`.

File: src/zarr/storage/common.py (hunk 1 above)
Comment (d-v-b) [reply src/zarr/storage/common.py:223]: good catch. I noticed that this wasn't failing any tests, so in addition to fixing this, I added tests for it in https://github.com/zarr-developers/zarr-python/pull/2384/commits/e6ecf99e67ef8e046214eeec297c014e367e59e0

File: src/zarr/storage/_utils.py [hunk 3]
```diff
@@ -1,11 +1,61 @@
 from __future__ import annotations
 
+from pathlib import Path
 from typing import TYPE_CHECKING
 
 if TYPE_CHECKING:
     from zarr.core.buffer import Buffer
 
 
+def normalize_path(path: str | bytes | Path | None) -> str:
```
Comment (TomAugspurger) [inline src/zarr/storage/_utils.py:10]: Does this need to take `| None` still? Or have we replaced all those with `""` as the default?

File: src/zarr/storage/_utils.py [hunk 4]
```diff
@@ -1,11 +1,61 @@
...
+
+    # convert backslash to forward slash
+    result = result.replace("\\", "/")
+
+    # ensure no leading slash
+    while len(result) > 0 and result[0] == "/":
+        result = result[1:]
+
+    # ensure no trailing slash
+    while len(result) > 0 and result[-1] == "/":
```
Comment (TomAugspurger) [inline src/zarr/storage/_utils.py:35]: Same question, but for `.rstrip`.

File: src/zarr/storage/_utils.py [hunk 5]
```diff
@@ -1,11 +1,61 @@
...
+        result = result[:-1]
+
+    # collapse any repeated slashes
```
Comment (TomAugspurger) [inline src/zarr/storage/_utils.py:38]: Maybe a regex is clearer?

```
pat = re.compile(r"//+")
result = pat.sub("/", result)
```

File: src/zarr/storage/_utils.py (hunk 3 above)
Comment (d-v-b) [reply src/zarr/storage/_utils.py:10]: I have not changed any of the top-level defaults, so this still needs to handle `None`.

File: src/zarr/storage/_utils.py (hunk 4 above)
Comment (d-v-b) [reply src/zarr/storage/_utils.py]: `strip` took care of both of them

File: src/zarr/storage/_utils.py (hunk 5 above)
Comment (d-v-b) [reply src/zarr/storage/_utils.py:33]: done

**Review by TomAugspurger — APPROVED** [review]

**Pull request merged**